File: src/BalancedBinaryTree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        if (getHeight(root) < 0) return false;
        return true;
    }
    
    public int getHeight(TreeNode root){
        if (root == null) return 0;
        int leftHeight = getHeight(root.left);
        if (leftHeight < 0) return -1;
        int rightHeight = getHeight(root.right);
        if (rightHeight < 0) return -1;
        if (Math.abs(leftHeight - rightHeight) > 1) return -1;
        return Math.max(leftHeight, rightHeight) + 1;
    }
}

File: BalancedBinaryTree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        initHeight(root);
        return isB(root);
    }
    
    bool isB(TreeNode *root){
        if (!root) return true;
        if (abs(getHeight(root->left) - getHeight(root->right)) <= 1 && isB(root->left) && isB(root->right)) return true;
        return false;
    }
    
    int getHeight(TreeNode *e){
        if (!e) return 0;
        return e->val;
    }
    
    int initHeight(TreeNode *root) {
        if (!root) return 0;
        root->val = max(initHeight(root->left), initHeight(root->right)) + 1;
        return root->val;
    }
};
